### ISSUE TITLE
Erratic normals on CSGPolygon when SmoothFaces is true

### ISSUE THREAD

**Issue by Windfisch** [godotengine/godot#22881]
**Godot version:** commit da6646dd5e5c779d72d5819bffce4a4e96e94344

**OS/device including version:** Linux/X11

**Issue description:**
When extruding a Polygon along a Path using a CSGPolygon and checking the "Smooth" checkbox, the normals get messed up.

**Actual behavior**:
When enabling "Smooth" on a square extruded along a straight path, this happens:

The below screenshots show the view space normals (i.e. `ALBEDO = NORMAL`)

![bug1](https://user-images.githubusercontent.com/4471484/46669092-b7abef00-cbce-11e8-949c-e3c67bb29765.png)

becomes

![bug2](https://user-images.githubusercontent.com/4471484/46669107-c1cded80-cbce-11e8-88f9-4e0224b16826.png)

**Expected behavior**:

Actual smooth normals, like the section in the left part of the image

**Steps to reproduce:**

1. Create a Path representing a straight line
2. Create a CSGPolygon using that path
3. Toggle "Smooth Faces"

**Minimal reproduction project:**

[bug.tar.gz](https://github.com/godotengine/godot/files/2460348/bug.tar.gz)

This may be related to https://github.com/godotengine/godot/issues/19916?

**Comment by Windfisch** [comment]
In `csg_shape.cpp`, in `CSGShape::_update_shape()`, smooth normals are calculated using the `OAHashMap<Vector3, Vector3> vec_map;` by iterating through all faces and their vertices `v`, looking up `v` in `vec_map` and adding the face normal to that `v`'s normal.

I suspect that due to rounding errors, this hashmap lookup sometimes fails to treat ("almost") equal vertices as equal sometimes, and thus assigns them distinct, unsmoothened normals. These rounding errors would also explain the pseudorandom pattern.

If that is the case, possible fixes would be either to avoid these rounding errors beforehand, or to use a fundamentally different approach using indexed vertex lists. That would ensure the floats to be exactly equal (or enable one to compare the indices instead of the float values in the first place)

**Comment by Windfisch** [comment]
I tracked this down to a bug in `OAHashMap`. It seems to lose elements.

A fix (together with lots of debugging output to illustrate the problem) can be found at https://github.com/Windfisch/godot/tree/hashmap_fixed . It exchanges the use of `OAHashMap<Vector3,Vector3>` by `HashMap<Vector3,Vector3>`.

Find attached the example to whose output I am referring to: [bug2.tar.gz](https://github.com/godotengine/godot/files/2465785/bug2.tar.gz)

With `OAHashMap`, I'm getting this:
```
	looking up 8.875809 -1.482800 -4.560672	looking up 410e0350 bfbdcc65 c091f106 -> miss
[...]
	looking up 8.875809 -1.482800 -4.560672	looking up 410e0350 bfbdcc65 c091f106 -> hit
[...]
	looking up 8.875809 -1.482800 -4.560672	looking up 410e0350 bfbdcc65 c091f106 -> hit
[... doing something else ...]
THIS SHOULD NOT HAPPEN, could not find 8.875809 -1.482800 -4.560672 = 410e0350 bfbdcc65 c091f106 at 369 2
```
The first miss is fine, the following hits are fine too. The last line indicates that `OAHashMap` had previously had an entry for *8.875809 -1.482800 -4.560672*, but now it can't find it any more. Despite it was never removed!


If you want, I can create a pull request from a cleaned up version of my fix; however, I'm not sure if there's a good reason to prefer `OAHashMap` instead. And obviously, the `OAHashMap` needs to be fixed.

With the `HashMap` fix, the last line does not show up, and the normals seem correct.

**Comment by Windfisch** [comment]
Yes, this most likely caused by issue #22928.

**Comment by akien-mga** [comment]
I assume it's fixed by #22929 then, thanks for the debugging and fix!
